sklearn_numba_dpex/kmeans/drivers.py

```diff
@@ -784,6 +793,211 @@ def _get_euclidean_distances(self, X, Y):
+        )
+
+        candidate_ids = dpt.empty(
+            sh=(n_local_trials,), dtype=np.int32, device=self.device
+        )
+
+        # Pick first center randomly
+        starting_center_id = np.int32(get_random_raw(random_state) % n_samples)
+        center_indices[0] = starting_center_id[0]
+
+        # track index of point, initialize list of closest distances and calculate
+        # current potential
+        kmeansplusplus_init_kernel(
+            X_t,
+            sample_weight,
+            centers_t,
+            center_indices,
+            closest_dist_sq,
+        )
+        total_potential = reduce_potential_1d_kernel(closest_dist_sq)
+
+        # Pick the remaining n_clusters-1 points
+        for c in range(1, n_clusters):
+            # NB: this step consists in highly sequential instructions that can only
+            # be parallelized in n_local_trials threads, it's up to 50x faster to run it
+            # on CPU, and becomes the bottleneck if ran on GPU. Depending on hardware
+            # and weight of data transfers, are there cases where it would be more
+            # efficient to keep it on GPU ?
+            # ???: would it be bad to sample the sample weight once outside the loop and
+            # reuse it at each iteration ?
+            if from_cpu_to_device:
+                candidate_ids = candidate_ids.to_device(cpu_device)
+                sample_center_candidates_kernel(
+                    closest_dist_sq.to_device(cpu_device),
+                    total_potential.to_device(cpu_device),
+                    random_state,
+                    candidate_ids,
+                )
+                candidate_ids = candidate_ids.to_device(self.device)
+            else:
+                sample_center_candidates_kernel(
+                    closest_dist_sq, total_potential, random_state, candidate_ids
+                )
+
+            # XXX: at the cost of one additional pass on data, we could avoid storing
+            # entirely distance_to_candidates_t in memory, and save
+            # `dtype.nbytes * n_local_trials * n_sample` bytes in memory.
+            # Which is better ?
+            kmeansplusplus_single_step_fixed_window_kernel(
+                X_t,
+                sample_weight,
+                candidate_ids,
+                closest_dist_sq,
+                distance_to_candidates_t,
+            )
+
+            candidate_potentials = reduce_potential_2d_kernel(distance_to_candidates_t)[
+                :, 0
+            ]
+            best_candidate = select_best_candidate_kernel(candidate_potentials)[0]
+
+            total_potential = candidate_potentials[
+                best_candidate : (best_candidate + 1)
+            ]
+            closest_dist_sq = distance_to_candidates_t[best_candidate]
+            best_candidate = candidate_ids[best_candidate]
+            centers_t[:, c] = X_t[:, best_candidate]
+            center_indices[c] = best_candidate
```

Comment:
To use two different variables for those two index while preserving the logic.
```suggestion
            center_index = candidate_ids[best_candidate]
            centers_t[:, c] = X_t[:, center_index]
            center_indices[c] = center_index
```